src/crop-manager.sol

```diff
@@ -0,0 +1,171 @@
+    }
+
+    function migrate(address newOwner) external {
+        require(msg.sender == owner, "UrnProxy/not-owner");
+        VatLike(vat).hope(newOwner);
+        VatLike(vat).nope(owner);
+        owner = newOwner;
+    }
+}
+
+contract CropJoinManager {
+    mapping (address => uint256) public wards;
+    mapping (address => address) public proxy;  // UrnProxy per user
+    address public implementation;
+
+    event Rely(address indexed);
+    event Deny(address indexed);
+    event SetImplementation(address indexed);
+
+    constructor() public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function rely(address usr) public auth { wards[usr] = 1; emit Rely(msg.sender); }
+    function deny(address usr) public auth { wards[usr] = 0; emit Deny(msg.sender); }
+    modifier auth { require(wards[msg.sender] == 1, "CropJoinManager/non-authed"); _; }
+
+
+    function setImplementation(address implementation_) external auth {
+        implementation = implementation_;
+        emit SetImplementation(implementation_);
+    }
+
+    fallback() external {
+        address _impl = implementation;
+        require(_impl != address(0));
+
+        assembly {
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize())
+            let result := delegatecall(gas(), _impl, ptr, calldatasize(), 0, 0)
+            let size := returndatasize()
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+}
+
+contract CropJoinManagerImp {
+    bytes32 slot0;
+    mapping (address => address) public proxy;  // UrnProxy per user
+
+    address public immutable vat;
+    constructor(address vat_) public {
+        vat = vat_;
+    }
+
+    function getProxy(address usr) internal returns (address urp) {
+        urp = proxy[usr];
+        if (urp == address(0)) {
+            urp = proxy[usr] = address(new UrnProxy(address(vat), usr));
+        }
+    }
+
+    function createProxy() external returns (address) {
+        return getProxy(msg.sender);
+    }
+
+    function join(address crop, address urn, uint256 val) external {
+        TokenLike(CropLike(crop).gem()).transferFrom(msg.sender, address(this), val);
+        TokenLike(CropLike(crop).gem()).approve(crop, val);
+        CropLike(crop).join(getProxy(urn), urn, val);
+    }
+
+    function exit(address crop, address usr, uint256 val) external {
+        CropLike(crop).exit(getProxy(msg.sender), usr, val);
+    }
+
+    function flee(address crop) external {
+        CropLike(crop).flee(getProxy(msg.sender));
+    }
+
+    function frob(address crop, address u, address v, address w, int256 dink, int256 dart) external {
+       require(u == msg.sender && v == msg.sender && w == msg.sender, "CropJoinManager/not-allowed");
+
+        VatLike(vat).frob(CropLike(crop).ilk(), getProxy(u), getProxy(v), w, dink, dart);
```

Comment:
nit: use a single `getProxy` call and cache result in a stack variable to save gas.

Creating the proxy seems pointless here as well--gotta `join` before you can `frob`.

And thirdly, it would be more efficient to just pass `ilk` here directly, rather than the CropJoin address

Reply:
Conclusion: add `ilk` to the extended `vat` interfaces

Reply:
Done.